### ISSUE TITLE
RuntimeError "can't add a new key into hash during iteration"

### ISSUE THREAD

**Issue by micdah** [fluent-plugins-nursery/fluent-plugin-concat#49]
#### Problem

I am seeing occasional warnings with the following log output
```
2018-06-12 09:11:49 +0000 [warn]: #0 dump an error event: error_class=RuntimeError error="can't add a new key into hash during iteration" location="/fluentd/vendor/bundle/ruby/2.3.0/gems/fluent-plugin-concat-2.2.2/lib/fluent/plugin/filter_concat.rb:122:in `process'" tag="kubernetes.var.log.containers.REDACTED.log" time=2018-06-12 09:11:48.674985171 +0000 record={"log"=>"REDACTED", "stream"=>"stdout"}
```

It could look like some kind of time-related issue, as I am not seeing an overwhelming amount of these.

#### Your environment

* OS version: Running inside kubernetes cluster on Azure AKS built based on image https://github.com/fluent/fluentd-kubernetes-daemonset tagged `v1.1-debian-elasticsearch`
```
$ uname -a
Linux fluentd-2wm7d 4.13.0-1018-azure #21-Ubuntu SMP Thu May 17 13:58:38 UTC 2018 x86_64 GNU/Linux
```
* paste result of ``fluentd --version`` or ``td-agent --version``
```
$ fluentd --version
fluentd 1.1.3
``` 
* plugin version
  * paste boot log of fluentd or td-agent
  * paste result of ``fluent-gem list``, ``td-agent-gem list`` or your Gemfile.lock
```
$ fluent-gem list
*** LOCAL GEMS ***

activesupport (5.2.0)
addressable (2.5.2)
azure-loganalytics-datacollector-api (0.1.2)
bigdecimal (1.2.8)
concurrent-ruby (1.0.5)
cool.io (1.5.3)
dig_rb (1.0.1)
domain_name (0.5.20180417)
elasticsearch (6.0.2)
elasticsearch-api (6.0.2)
elasticsearch-transport (6.0.2)
excon (0.62.0)
faraday (0.15.0)
ffi (1.9.23)
fluent-plugin-azure-loganalytics (0.3.1)
fluent-plugin-concat (2.2.2)
fluent-plugin-elasticsearch (2.9.2)
fluent-plugin-kubernetes_metadata_filter (2.1.2)
fluent-plugin-record-modifier (1.0.2)
fluent-plugin-systemd (1.0.0)
fluentd (1.1.3)
http (0.9.8)
http-cookie (1.0.3)
http-form_data (1.0.3)
http_parser.rb (0.6.0)
i18n (1.0.1)
io-console (0.4.5)
json (1.8.3)
kubeclient (1.1.4)
lru_redux (1.1.0)
mime-types (3.1)
mime-types-data (3.2016.0521)
minitest (5.11.3)
msgpack (1.2.4)
multi_json (1.13.1)
multipart-post (2.0.0)
netrc (0.11.0)
oj (3.5.1)
psych (2.1.0)
public_suffix (3.0.2)
rdoc (4.2.1)
recursive-open-struct (1.0.0)
rest-client (2.0.2)
serverengine (2.0.6)
sigdump (0.2.4)
strptime (0.2.3)
systemd-journal (1.3.1)
thread_safe (0.3.6)
tzinfo (1.2.5)
tzinfo-data (1.2018.5)
unf (0.1.4)
unf_ext (0.0.7.5)
yajl-ruby (1.4.0)
```

**Comment by okkez** [comment]
@micdah Could you try #50 ?

**Comment by micdah** [comment]
@okkez yes, but how would I go about trying it out? Can I just build a new docker image, where i overwrite the latest version of the plugin with the updated `.rb` file?

**Comment by micdah** [comment]
@okkez Alright, I downloaded the updated version and copied it into our docker image using
```
COPY ./filter_concat.rb /fluentd/vendor/bundle/ruby/2.3.0/gems/fluent-plugin-concat-2.2.2/lib/fluent/plugin/filter_concat.rb
```

I have had the new version running for now close to half an hour, and checking the logs I find no instances of this error anymore.

**Comment by okkez** [comment]
Thank you for checking my patch!
I will release new version as soon as possible.

**Comment by okkez** [comment]
I've released new version v2.2.3

**Comment by micdah** [comment]
@okkez Thank you for your quick reaction time - that is awesome! 👍
